elastic/kibana
# [RAC] Initializes Metrics and Logs rule registry setup

## :memo: Summary

This makes the metrics and logs alert rules life-cycle-aware.

## :detective: Review notes

- *Security solution* and *APM* teams: The changes in the types for the `createLifecycleRuleType` and `createPersistenceRuleType` helpers necessitated some type amendments in your code. Let me know whether those are acceptable.
- The `createLifecycleRuleType` and `createPersistenceRuleType` helpers now preserve all generics that the types of the `alerting` plugin expose so that strictly typed alert registrations can continue to make use of them.
- The client-side registration required for the alert table will be implemented separately.
- The metrics alerts now receive properly validated and typed `params`. This required some refinement of the validation, so some additional manual testing of the creation of those would be beneficial.
- The alerts currently don't persist any fields in the life-cycle document.

## Testing

To test these changes:
1. Switch to this branch
2. `yarn es snapshot --ssl --license=trial`
3. Run metricbeat locally with these configs in metricbeat.yml:

```yaml
output.elasticsearch:
  # Array of hosts to connect to.
  hosts: ["localhost:9200"]

  # Protocol - either `http` (default) or `https`.
  protocol: "https"
  ssl.verification_mode: "none"
```

To start, run:

```
./metricbeat setup
./metricbeat -e
```

The system module should be enabled.

4. Run Kibana (no need to run Kibana as HTTPS) but with these settings in kibana.dev.yml:

```yaml
elasticsearch.username: kibana
elasticsearch.password: changeme
elasticsearch.hosts:
  - https://localhost:9200
elasticsearch.ssl.verificationMode: none

xpack.observability.unsafe.alertingExperience.enabled: true
```

5. Open Kibana and go to the Metrics UI. There should be data flowing in, you should see your local node in the inventory view. Choose "Alerts > Metrics > Create threshold alert" in the top right menu. 
6. Choose a value that will alert a lot (and keep the "Check every" low, to 1 min is fine)

<img width="573" alt="Screen Shot 2021-06-04 at 11 41 38 AM" src="https://user-images.githubusercontent.com/159370/120827885-eb7b7600-c529-11eb-914f-fc300ddf99cd.png">

7. Connect it to a Server Log if you want to verify it's running.
8. Check your Kibana logs to watch for the first server log to emit the alert message.
9. Navigate to /app/observability/alerts and you should see the alerts table, with (at least) 1 alert, like this:

<img width="1895" alt="Screen Shot 2021-06-03 at 9 15 27 PM" src="https://user-images.githubusercontent.com/159370/120828108-22518c00-c52a-11eb-8201-b74d6cb19c97.png">

NOTE: Because we haven't implemented a custom reason or a "view in app" link, we are missing lots of parts of each row here, but it's showing up! Also, because [the underlying rule registry is not quite working yet](https://github.com/elastic/kibana/issues/100451), there is a new row for every evaluation instead of just one row for the ongoing alert, which is what it will do before 7.14, hopefully. 

## Original WIP notes by @jasonrhodes:

* I had to move the executor inline into the register file because I couldn't otherwise figure out how to split out the type that would correspond to the options given to the executor from the registry plugin. By doing it inline, I don't have to specify those options, and the types work well.
* There is a "@ts-expect-error" that needs to be dealt with. We were doing a few "as" casts that made me nervous so I'd like to figure those out.
* The typescript magic here is amazing but also I don't understand how it works :P
* I was largely [looking at this APM example](https://github.com/elastic/kibana/blob/bdde884d098f42a93630cc6d221258bdc3372e60/x-pack/plugins/apm/server/lib/alerts/register_error_count_alert_type.ts#L51), and I don't understand how they are registering a rule without specifying their action group IDs the way we were specifying ours in our own types, but maybe that wasn't necessary/was overkill? Our types for the registration method have been removed and we are relying on type inference now, and it seems to work?
* We still need to implement our "custom reason" message and our "view in app" link for this (and every other) rule type
